### ISSUE TITLE
ValueError: Unable to process pseudo var AWS::StackId

### ISSUE THREAD

**Issue by xentripetal** [DontShaveTheYak/cf2tf#22]
Thanks for the tool! I was testing it out for [this template](https://aws-neptune-customer-samples.s3.amazonaws.com/neptune-stream/neptune_to_elastic_search.json)

which gives the error

```bash
$ cf2tf neptune_to_elastic_search.json > test.tf
// Converting neptune_to_elastic_search.json to Terraform!
Traceback (most recent call last):
  File "/opt/homebrew/bin/cf2tf", line 8, in <module>
    sys.exit(cli())
  File "/opt/homebrew/lib/python3.9/site-packages/click/core.py", line 1130, in __call__
    return self.main(*args, **kwargs)
  File "/opt/homebrew/lib/python3.9/site-packages/click/core.py", line 1055, in main
    rv = self.invoke(ctx)
  File "/opt/homebrew/lib/python3.9/site-packages/click/core.py", line 1404, in invoke
    return ctx.invoke(self.callback, **ctx.params)
  File "/opt/homebrew/lib/python3.9/site-packages/click/core.py", line 760, in invoke
    return __callback(*args, **kwargs)
  File "/opt/homebrew/lib/python3.9/site-packages/cf2tf/app.py", line 44, in cli
    config = TemplateConverter(cf_template, search_manger).convert()
  File "/opt/homebrew/lib/python3.9/site-packages/cf2tf/convert.py", line 59, in convert
    tf_resources = self.convert_to_tf(self.manifest)
  File "/opt/homebrew/lib/python3.9/site-packages/cf2tf/convert.py", line 111, in convert_to_tf
    tf_resources.extend(converter(resources))
  File "/opt/homebrew/lib/python3.9/site-packages/cf2tf/convert.py", line 226, in convert_conditions
    resolved_values = self.resolve_values(dict_conditons, functions.ALL_FUNCTIONS)
  File "/opt/homebrew/lib/python3.9/site-packages/cf2tf/convert.py", line 138, in resolve_values
    data[key] = self.resolve_values(value, allowed_func)
  File "/opt/homebrew/lib/python3.9/site-packages/cf2tf/convert.py", line 144, in resolve_values
    value = self.resolve_values(value, functions.ALLOWED_FUNCTIONS[key])
  File "/opt/homebrew/lib/python3.9/site-packages/cf2tf/convert.py", line 150, in resolve_values
    return [self.resolve_values(item, allowed_func) for item in data]
  File "/opt/homebrew/lib/python3.9/site-packages/cf2tf/convert.py", line 150, in <listcomp>
    return [self.resolve_values(item, allowed_func) for item in data]
  File "/opt/homebrew/lib/python3.9/site-packages/cf2tf/convert.py", line 142, in resolve_values
    raise ValueError(f"{key} not allowed here.")
ValueError: Fn::Join not allowed here.
```

A [related template](https://raw.githubusercontent.com/awslabs/amazon-neptune-tools/master/export-neptune-to-elasticsearch/cloudformation-templates/export-neptune-to-elasticsearch.json) gives a similar error where it can't handle different data types on join parameters 

```bash
$ cf2tf export.json > another_test.tf
// Converting export.json to Terraform!
Traceback (most recent call last):
  File "/opt/homebrew/bin/cf2tf", line 8, in <module>
    sys.exit(cli())
  File "/opt/homebrew/lib/python3.9/site-packages/click/core.py", line 1130, in __call__
    return self.main(*args, **kwargs)
  File "/opt/homebrew/lib/python3.9/site-packages/click/core.py", line 1055, in main
    rv = self.invoke(ctx)
  File "/opt/homebrew/lib/python3.9/site-packages/click/core.py", line 1404, in invoke
    return ctx.invoke(self.callback, **ctx.params)
  File "/opt/homebrew/lib/python3.9/site-packages/click/core.py", line 760, in invoke
    return __callback(*args, **kwargs)
  File "/opt/homebrew/lib/python3.9/site-packages/cf2tf/app.py", line 38, in cli
    cf_template = Template.from_yaml(tmpl_path).template
  File "/opt/homebrew/lib/python3.9/site-packages/cf2tf/cloudformation/_template.py", line 77, in from_yaml
    tmp_yaml = load_yaml(raw)
  File "/opt/homebrew/lib/python3.9/site-packages/cfn_tools/__init__.py", line 29, in load_yaml
    return yaml.load(source, Loader=CfnYamlLoader)
  File "/opt/homebrew/lib/python3.9/site-packages/yaml/__init__.py", line 81, in load
    return loader.get_single_data()
  File "/opt/homebrew/lib/python3.9/site-packages/yaml/constructor.py", line 49, in get_single_data
    node = self.get_single_node()
  File "/opt/homebrew/lib/python3.9/site-packages/yaml/composer.py", line 36, in get_single_node
    document = self.compose_document()
  File "/opt/homebrew/lib/python3.9/site-packages/yaml/composer.py", line 55, in compose_document
    node = self.compose_node(None, None)
  File "/opt/homebrew/lib/python3.9/site-packages/yaml/composer.py", line 84, in compose_node
    node = self.compose_mapping_node(anchor)
  File "/opt/homebrew/lib/python3.9/site-packages/yaml/composer.py", line 133, in compose_mapping_node
    item_value = self.compose_node(node, item_key)
  File "/opt/homebrew/lib/python3.9/site-packages/yaml/composer.py", line 84, in compose_node
    node = self.compose_mapping_node(anchor)
  File "/opt/homebrew/lib/python3.9/site-packages/yaml/composer.py", line 133, in compose_mapping_node
    item_value = self.compose_node(node, item_key)
  File "/opt/homebrew/lib/python3.9/site-packages/yaml/composer.py", line 84, in compose_node
    node = self.compose_mapping_node(anchor)
  File "/opt/homebrew/lib/python3.9/site-packages/yaml/composer.py", line 133, in compose_mapping_node
    item_value = self.compose_node(node, item_key)
  File "/opt/homebrew/lib/python3.9/site-packages/yaml/composer.py", line 84, in compose_node
    node = self.compose_mapping_node(anchor)
  File "/opt/homebrew/lib/python3.9/site-packages/yaml/composer.py", line 133, in compose_mapping_node
    item_value = self.compose_node(node, item_key)
  File "/opt/homebrew/lib/python3.9/site-packages/yaml/composer.py", line 84, in compose_node
    node = self.compose_mapping_node(anchor)
  File "/opt/homebrew/lib/python3.9/site-packages/yaml/composer.py", line 133, in compose_mapping_node
    item_value = self.compose_node(node, item_key)
  File "/opt/homebrew/lib/python3.9/site-packages/yaml/composer.py", line 84, in compose_node
    node = self.compose_mapping_node(anchor)
  File "/opt/homebrew/lib/python3.9/site-packages/yaml/composer.py", line 133, in compose_mapping_node
    item_value = self.compose_node(node, item_key)
  File "/opt/homebrew/lib/python3.9/site-packages/yaml/composer.py", line 82, in compose_node
    node = self.compose_sequence_node(anchor)
  File "/opt/homebrew/lib/python3.9/site-packages/yaml/composer.py", line 111, in compose_sequence_node
    node.value.append(self.compose_node(node, index))
  File "/opt/homebrew/lib/python3.9/site-packages/yaml/composer.py", line 82, in compose_node
    node = self.compose_sequence_node(anchor)
  File "/opt/homebrew/lib/python3.9/site-packages/yaml/composer.py", line 110, in compose_sequence_node
    while not self.check_event(SequenceEndEvent):
  File "/opt/homebrew/lib/python3.9/site-packages/yaml/parser.py", line 98, in check_event
    self.current_event = self.state()
  File "/opt/homebrew/lib/python3.9/site-packages/yaml/parser.py", line 486, in parse_flow_sequence_entry
    if self.check_token(KeyToken):
  File "/opt/homebrew/lib/python3.9/site-packages/yaml/scanner.py", line 116, in check_token
    self.fetch_more_tokens()
  File "/opt/homebrew/lib/python3.9/site-packages/yaml/scanner.py", line 258, in fetch_more_tokens
    raise ScannerError("while scanning for the next token", None,
yaml.scanner.ScannerError: while scanning for the next token
found character '\t' that cannot start any token
  in "<unicode string>", line 344, column 1:
    		"release_2022_06_06",
    ^
```

No urgency in investigating this, just figured I'd report these issues for future hardening.

**Comment by shadycuz** [comment]
Will take a look at it. Thanks for letting me know.

**Comment by shadycuz** [comment]
@xentripetal I believe the first issue is a bad template. AWS [documentation ](https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/intrinsic-function-reference-conditions.html#w2ab1c31c28c21c29)clearly states that the conditional function `Fn::Equals` may have nested inside it `Fn::FindInMap`, `Ref` or other condition functions. Since `Fn::Join` is an intrinsic function the error `ValueError: Fn::Join not allowed here.` is correct. Though the error message is not that great.

I have changed the error message to `ValueError: Fn::Join not allowed to be nested in Fn::Equals.`

Also, even if `Fn::Join` was allowed in `Fn::Equals`, the template still doesn't make much sense. 

```JSON
"Conditions": {
    "CreateManagedPolicy": {
      "Fn::Equals": [
        {
          "Fn::Join": [
            ",",
            {
              "Ref": "ManagedPolicies"
            }
          ]
        },
        ""
      ]
    },
```

CreateManagedPolicy will always be false because `Ref + "," will never == ""`


For the second issue, there is an illegal character in the file. 

```
found character '\t' that cannot start any token
  in "<unicode string>", line 344, column 1:
                "release_2022_06_06",
    ^
```

When I open it up locally I can see it. 

![image](https://user-images.githubusercontent.com/17090999/175837716-a69a2875-203e-4e1f-95e1-7c51ca86072c.png)

Unfortunately removing the illegal  character doesn't get us much further =/ 

``` console
File "/workspaces/cf2tf/src/cf2tf/conversion/expressions.py", line 809, in handle_pseduo_var
    raise ValueError(f"Unable to process pseudo var {pseudo_name}.")
ValueError: Unable to process pseudo var AWS::StackId.
```

This stack uses the AWS Pseudo var `StackId`. Their isn't really a Terraform equivalent and I'm not sure how to handle these. The `StackId` is [bascially a random string/number](https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/pseudo-parameter-reference.html#cfn-pseudo-param-stackid). 

Terraform does offer a [uuid](https://www.terraform.io/language/functions/uuid) and [uuidv5](https://www.terraform.io/language/functions/uuidv5) which could be a good alternative. Does replacing `StackId` with `uuidv5("foo", "bar")` sound like a good solution to you?

**Comment by xentripetal** [comment]
> I believe the first issue is a bad template

It does seem that the template violates the docs page. However it is an official template that launches successfully in CF.
https://docs.aws.amazon.com/neptune/latest/userguide/full-text-search-cfn-create.html

So I'm not sure what to make of that. Your call on whether to go by the documentation standards or implementation standards.

> Also, even if Fn::Join was allowed in Fn::Equals, the template still doesn't make much sense.

Its a little confusing but the ManagedPolicies reference is a list of strings and the first parameter to `Fn:Join` is a delimiter not a joining element. So its the equivalent of checking if the list is empty.

TF equivalent would be
```hcl
variable "managed_policies" {
  type    = list(string)
  default = []
}

locals (
    create_managed_policies = join("/", var.managed_policies) == ""
)
```

---

> For the second issue, there is an illegal character in the file.

Hmm, from what I could find JSON permits any whitespace  (tabulation (U+0009), line feed (U+000A), carriage return (U+000D), and space (U+0020)) before any token. Seems like that should be valid JSON. What are you using to parse it?

> Does replacing StackId with uuidv5("foo", "bar") sound like a good solution to you?

I think thats reasonable. It would definitely be a little confusing in AWS to see what appears to be a CF reference but have it not exist in CF. But from a translation perspective, thats definitely the most accurate behavior. I'd maybe consider defining a
```hcl
resource "random_uuid" "stack_id" {
}

random_uuid.stack_id.id
```
so it stays persistent across references.

---

On a related note to pseudo vars, it seems that AWS::Partition gets resolved to
```hcl
data "aws_region" "current" {
}

data.aws_region.current.name
```

instead of 
```
data "aws_partition" "current" {
}

data.aws_partition.current.name
```

when inside a Fn:Sub. 

e.g. https://s3.amazonaws.com/aws-neptune-customer-samples/neptune-stream/neptune_stream_poller_lambda_stack.json
```json
"Resource": {
  "Fn::Sub": "arn:${AWS::Partition}:lambda:*:*:function:*"
}
```

**Comment by shadycuz** [comment]
> However it is an official template that launches successfully in CF. https://docs.aws.amazon.com/neptune/latest/userguide/full-text-search-cfn-create.html

Yeah it looked like an AWS quickstart template since it was coming from an AWS bucket.

> So I'm not sure what to make of that. Your call on whether to go by the documentation standards or implementation standards.

Well my last AWS documentation [PR ](https://github.com/awsdocs/aws-cloudformation-user-guide/pull/1013) has been open for a year now...

So I guess we will just allow it.


> Its a little confusing but the ManagedPolicies reference is a list of strings and the first parameter to `Fn:Join` is a delimiter not a joining element. So its the equivalent of checking if the list is empty.

Yeah 😊 I see that now.

> Hmm, from what I could find JSON permits any whitespace (tabulation (U+0009), line feed (U+000A), carriage return (U+000D), and space (U+0020)) before any token. Seems like that should be valid JSON. What are you using to parse it?

I use an AWS library to parse it https://github.com/awslabs/aws-cfn-template-flip

> I think thats reasonable. It would definitely be a little confusing in AWS to see what appears to be a CF reference but have it not exist in CF. But from a translation perspective, thats definitely the most accurate behavior. I'd maybe consider defining a

I will add `StackId` and `StackName`, hopefully this weekend.


>On a related note to pseudo vars, it seems that AWS::Partition gets resolved to
>
> ```terraform
> data "aws_region" "current" {
> }
> 
> data.aws_region.current.name
> ```
> 
> instead of
> 
> ```
> data "aws_partition" "current" {
> }
> 
> data.aws_partition.current.name
> ```
> 
> when inside a Fn:Sub.
> 
> e.g. https://s3.amazonaws.com/aws-neptune-customer-samples/neptune-stream/neptune_stream_poller_lambda_stack.json
> 
> ```json
> "Resource": {
>   "Fn::Sub": "arn:${AWS::Partition}:lambda:*:*:function:*"
> }
> ```
hmmm Well this a bug 🐛

---

After I allowed join inside equals, I hit another error with the template.

```console
File "/workspaces/cf2tf/src/cf2tf/conversion/expressions.py", line 332, in find_in_map
    raise KeyError(f"Unable to find {map_name} in locals block.")
KeyError: 'Unable to find var.query_engine in locals block.'
```

for this code
```json
"Fn::FindInMap": [
  {
    "Ref": "QueryEngine"
  },
  {
    "Ref": "LambdaRuntime"
  },
  {
    "Ref": "EnableNonStringIndexing"
  }
]
```

What I do for Cloudformation `Mappings` section, is convert them to a local value:
```HCL
locals {
  foo = { bar = { baz = "foobar"} }
}
```

And then when we come across `FindInMap` with values `["foo", "bar", "baz"]` we turn it into `locals.foo["bar"]["baz"]`.

But in this template the map name is a variable. I'm not sure how to handle this in Terraform. I'm open to suggestions though.

**Comment by xentripetal** [comment]
>I'm open to suggestions though.

Hmm, it seems like theres an open request to add support for treating locals as a map which would make this behavior possible https://github.com/hashicorp/terraform/issues/27715

Excluding that though, I think the only option is to have a container local map that all maps go into. Its definitely ugly but it does support that behavior.

e.g.

```hcl
locals {
    cf_maps = { foo = { bar = { baz = "foobar}}}
}
```

However, I imagine templates using this behavior are likely rare and it would be fine to just leave a comment where the FindInMap call would go and leave it up the user to manually translate it.

**Comment by shadycuz** [comment]
@xentripetal I will push a new version soon but I have fixed the issues locally and converted your template.

```console
$ cf2tf neptune_to_elastic_search.json
// Converting neptune_to_elastic_search.json to Terraform!
 existing repo found.
```
```HCL
data "aws_region" "current" {
}


data "aws_partition" "current" {
}


locals  {
  mappings = {
    RuntimeMap = {
      python3.9 = {
        keyname = "python39/release_2022_06_06"
      }
      java8 = {
        keyname = "java8"
      }
    }
    Gremlin = {
      python3.9 = {
        true = "neptune_to_es.neptune_gremlin_es_handler.ElasticSearchGremlinHandler"
        false = "neptune_to_es.neptune_gremlin_es_string_indexing_handler.ElasticSearchStringOnlyGremlinHandler"
      }
      java8 = {
        true = "com.amazonaws.neptune.ElasticsearchGremlinReplicationHandler"
        false = "com.amazonaws.neptune.ElasticsearchGremlinReplicationHandler"
      }
    }
    Sparql = {
      python3.9 = {
        true = "neptune_to_es.neptune_sparql_es_handler.ElasticSearchSparqlHandler"
        false = "neptune_to_es.neptune_sparql_es_string_indexing_handler.ElasticSearchStringOnlySparqlHandler"
      }
      java8 = {
        true = "com.amazonaws.neptune.ElasticsearchSparqlReplicationHandler"
        false = "com.amazonaws.neptune.ElasticsearchSparqlReplicationHandler"
      }
    }
    ReplicationScopeMap = {
      All = {
        keyname = "all"
      }
      Nodes = {
        keyname = "nodes"
      }
    }
  }
  CreateManagedPolicy = join(",", var.managed_policies) == ""
  IsEmptyLambdaS3Bucket = var.lambda_s3_bucket == ""
  IsEmptyLambdaS3Key = var.lambda_s3_key == ""
  IsEmptyStreamRecordsHandler = var.stream_records_handler == ""
  CreateCloudWatchAlarmCondition = var.create_cloud_watch_alarm == "true"
}


variable "application_name" {
  description = "Application Name used as a reference to create resources"
  type = string
  default = "NeptuneStream"
}


variable "lambda_memory_size" {
  description = "Poller Lambda  memory size (in MB). Should be one of 128 MB, 256 MB, 512 MB, 1024 MB, 2048 MB."
  type = string
  default = "2048"
}


variable "lambda_runtime" {
  description = "Lambda Runtime"
  type = string
  default = "python3.9"
}


variable "lambda_s3_bucket" {
  description = "S3 bucket having Lambda Artifact. Optional Parameter - If left blank artifact will be picked from default Bucket."
  type = string
}


variable "lambda_s3_key" {
  description = "S3 key for Lambda Artifact. Optional Parameter - If left blank default Lambda Artifact is used."
  type = string
}


variable "lambda_logging_level" {
  description = "Poller Lambda logging level."
  type = string
  default = "INFO"
}


variable "managed_policies" {
  description = "Comma-delimited list of ARNs of managed policies to be attached to Lambda execution role. Optional Parameter - If left blank policy with required access is created. "
  type = string
}


variable "stream_records_handler" {
  description = "Handler for processing stream records. Optional Parameter - If left blank default Handler for Elastic Search is used."
  type = string
}


variable "stream_records_batch_size" {
  description = "Number of records to be read from stream in each batch. Should be between 1 to 50000."
  type = string
  default = "5000"
}


variable "max_polling_wait_time" {
  description = "Maximum wait time in seconds between two successive polling from stream. Set value to 0 sec for continuous polling. Maximum value can be 3600 sec (1 hour)."
  type = string
  default = "60"
}


variable "max_polling_interval" {
  description = "Period for which we can continuously poll stream for records on one Lambda instance. Should be between 5 sec to 900 sec. This parameter is used to set Poller Lambda Timeout."
  type = string
  default = "600"
}


variable "neptune_stream_endpoint" {
  description = "Endpoint for source Neptune Stream. This is of the form http(s)://<cluster>:<port>/gremlin/stream or http(s)://<cluster>:<port>/sparql/stream."
  type = string
}


variable "query_engine" {
  description = "Neptune Query Engine."
  type = string
  default = "Gremlin"
}


variable "iam_auth_enabled_on_source_stream" {
  description = "Flag to determine if IAM Auth is Enabled for Source Neptune Cluster or not."
  type = string
  default = "false"
}


variable "stream_db_cluster_resource_id" {
  description = "Neptune DB Cluster Resource Id. Ex: cluster-5DSWZGISGVCHJPHOV5MK7QF2PY. Optional Parameter- Only needed when IAM Auth is Enabled."
  type = string
}


variable "step_function_fallback_period" {
  description = "Period after which Step function is invoked using Cloud Watch Events to recover from failure. Unit for Step Function Fallback period is set separately."
  type = string
  default = "5"
}


variable "step_function_fallback_period_unit" {
  description = "Step Function FallbackPeriod unit. Should be one of minutes, minute, hours, hour, days, day"
  type = string
  default = "minutes"
}


variable "notification_sns_topic_arn" {
  description = "SNS Topic ARN where CloudWatch Alarm Notifications would be sent. Eg. arn:aws:sns:<region>:<account-id>:<name>. Optional."
  type = string
}


variable "notification_email" {
  description = "Email Address for CloudWatch Alarm Notification. Optional Parameter - Only needed when selecting option to create CloudWatch Alarm."
  type = string
}


variable "vpc" {
  description = "The VPC in which Neptune Stream Instance is present"
  type = string
}


variable "subnet_ids" {
  description = "The subnets to which a network interface is established. Add subnets associated with both Neptune Stream Cluster & Neptune target Cluster."
  type = string
}


variable "security_group_ids" {
  description = "The Security groups associated with the Neptune Stream Cluster and Neptune Target Cluster."
  type = string
}


variable "route_table_ids" {
  description = "Comma Delimited list of Route table ids associated with the Subnets. For Example: rtb-a12345,rtba7863k1. Optional parameter - Only needed when creating DynamoDB VPC Endpoint."
  type = string
}


variable "create_ddbvpc_end_point" {
  description = "Flag used to determine whether to create Dynamo DB VPC Endpoint or not. Select false only if Dynamo DB VPC endpoint already present."
  type = string
  default = "true"
}


variable "create_monitoring_end_point" {
  description = "Flag used to determine whether to create Monitoring VPC Endpoint or not. Select false only if Monitoring VPC endpoint already present."
  type = string
  default = "true"
}


variable "create_cloud_watch_alarm" {
  description = "Flag used to determine whether to create Cloud watch alarm or not."
  type = string
  default = "false"
}


variable "elastic_search_endpoint" {
  description = "Elastic Search Cluster Endpoint. Ex : vpc-neptunestream.us-east-1.es.amazonaws.com"
  type = string
}


variable "number_of_shards" {
  description = "Number of Shards for Elastic Search Index. Default value is 5."
  type = string
  default = "5"
}


variable "number_of_replica" {
  description = "Number of replicas for Elastic Search Index. Default value is 1."
  type = string
  default = "1"
}


variable "geo_location_fields" {
  description = "Comma Delimited list of Property Keys to be mapped to Geo Point Type in Elastic Search. For Example: location,area. Currently, for a field to be mapped to Geo Point type, value should be in the format 'latitude,longitude' Ex: '41.33,-11.69'"
  type = string
}


variable "properties_to_exclude" {
  description = "Comma delimited list of Property Keys to exclude from being indexed into Elastic Search. Optional Parameter - If left blank, all property keys will be indexed."
  type = string
}


variable "datatypes_to_exclude" {
  description = "Comma delimited list of Property Value Data Types to exclude from being indexed into Elastic Search. Optional Parameter - If left blank, all valid property values will be indexed. Type inputs that are unsupported for the specified query language will be ignored. Valid inputs for Sparql/RDF data: [string, boolean, float, double, dateTime, byte, int, long, short, date, decimal, integer, nonNegativeInteger, nonPositiveInteger, negativeInteger, unsignedByte, unsignedInt, unsignedLong, unsignedShort, time]. Valid inputs for Gremlin data: [string, date, bool, byte, short, int, long, float, double]"
  type = string
}


variable "enable_non_string_indexing" {
  description = "Flag to enable/disable indexing Non-String fields"
  type = string
  default = "true"
}


variable "replication_scope" {
  description = "Determines whether to replicate both nodes and edges, or only nodes (Gremlin engine only)."
  type = string
  default = "All"
}


variable "ignore_missing_document" {
  description = "Flag to determine if missing document error in Elastic Search can be ignored. Missing document error can occur rarely but will need manual intervention if not ignored."
  type = string
  default = "true"
}


resource "aws_iam_policy" "managed_policy" {
  description = "Policy for Elastic Search Access for Neptune Lambda Poller"
  path = "/"
  policy = {
    Version = "2012-10-17"
    Statement = [{'Action': ['"es:ESHttpDelete"', '"es:ESHttpGet"', '"es:ESHttpHead"', '"es:ESHttpPost"', '"es:ESHttpPut"'], 'Effect': '"Allow"', 'Resource': ['"arn:${data.aws_region.current.name}:es:*:*:*"']}]
  }
}


resource "aws_cloudformation_stack" "neptune_stream_poller" {
  parameters = {
    AdditionalParams = "{ "ElasticSearchEndpoint": "${var.elastic_search_endpoint}", "NumberOfShards": "${var.number_of_shards}", "NumberOfReplica": "${var.number_of_replica}", "IgnoreMissingDocument": "${var.ignore_missing_document}", "ReplicationScope": "${local.mappings["ReplicationScopeMap"][var.replication_scope]["keyname"]}", "GeoLocationFields": "${var.geo_location_fields}", "DatatypesToExclude": "${var.datatypes_to_exclude}", "PropertiesToExclude": "${var.properties_to_exclude}", "EnableNonStringIndexing": "${var.enable_non_string_indexing}"}"
    ApplicationName = var.application_name
    LambdaMemorySize = var.lambda_memory_size
    LambdaRuntime = var.lambda_runtime
    LambdaS3Bucket = local.IsEmptyLambdaS3Bucket ? join("-", ["aws-neptune-customer-samples", data.aws_region.current.name]) : var.lambda_s3_bucket
    LambdaS3Key = local.IsEmptyLambdaS3Key ? join("/", ["neptune-stream", "lambda", local.mappings["RuntimeMap"][var.lambda_runtime]["keyname"], "neptune-to-es.zip"]) : var.lambda_s3_key
    ManagedPolicies = local.CreateManagedPolicy ? aws_iam_policy.managed_policy.id : join(",", var.managed_policies)
    LambdaLoggingLevel = var.lambda_logging_level
    StreamRecordsHandler = local.IsEmptyStreamRecordsHandler ? local.mappings[var.query_engine][var.lambda_runtime][var.enable_non_string_indexing] : var.stream_records_handler
    StreamRecordsBatchSize = var.stream_records_batch_size
    StepFunctionFallbackPeriod = var.step_function_fallback_period
    StepFunctionFallbackPeriodUnit = var.step_function_fallback_period_unit
    MaxPollingWaitTime = var.max_polling_wait_time
    NeptuneStreamEndpoint = var.neptune_stream_endpoint
    IAMAuthEnabledOnSourceStream = var.iam_auth_enabled_on_source_stream
    StreamDBClusterResourceId = var.stream_db_cluster_resource_id
    MaxPollingInterval = var.max_polling_interval
    VPC = var.vpc
    RouteTableIds = var.route_table_ids
    CreateDDBVPCEndPoint = var.create_ddbvpc_end_point
    CreateMonitoringEndPoint = var.create_monitoring_end_point
    CreateCloudWatchAlarm = var.create_cloud_watch_alarm
    NotificationSNSTopicArn = var.notification_sns_topic_arn
    NotificationEmail = var.notification_email
    SubnetIds = join(",", var.subnet_ids)
    SecurityGroupIds = join(",", var.security_group_ids)
  }
  template_url = "https://s3.amazonaws.com/aws-neptune-customer-samples/neptune-stream/neptune_stream_poller_nested_full_stack.json"
}


output "https_access_sg" {
  description = "HTTPS Access Security Group Arn"
  value = aws_cloudformation_stack.neptune_stream_poller.outputs.HTTPSAccessSG
}


output "lease_dynamo_db_table" {
  description = "Neptune Stream Poller Lease Table"
  value = aws_cloudformation_stack.neptune_stream_poller.outputs.LeaseDynamoDBTable
}


output "state_machine_arn" {
  description = "Neptune Stream Poller State Machine Arn"
  value = aws_cloudformation_stack.neptune_stream_poller.outputs.StateMachineArn
}


output "cron_arn" {
  description = "Neptune Stream Poller Scheduler Cron Arn"
  value = aws_cloudformation_stack.neptune_stream_poller.outputs.CronArn
}


output "state_machine_alarm_arn" {
  description = "Neptune Stream Poller State Machine Alarm Arn"
  value = aws_cloudformation_stack.neptune_stream_poller.outputs.StateMachineAlarmArn
}


output "neptune_stream_poller_lambda_arn" {
  description = "Neptune Stream Poller Lambda Arn"
  value = aws_cloudformation_stack.neptune_stream_poller.outputs.NeptuneStreamPollerLambdaArn
}


output "cloud_watch_metrics_dashboard_uri" {
  description = "CloudWatch Metrics Dashboard URI"
  value = aws_cloudformation_stack.neptune_stream_poller.outputs.CloudWatchMetricsDashboardURI
}
```

**Comment by shadycuz** [comment]
[v0.2.3](https://github.com/DontShaveTheYak/cf2tf/releases/tag/0.2.3) is out! All of these bugs except the StackId have been fixed! Many thanks @xentripetal 

`pip install -U cf2tf` 🚀

**Comment by arroadie** [comment]
Just to notify that this also happens with [this template](https://aws-quickstart.s3.amazonaws.com/quickstart-apache-superset/templates/superset-entrypoint-existing-vpc.template.yaml)

**Comment by shadycuz** [comment]
Many weekends later and I finally got StackId and StackName implemented in `0.3.0`.
